tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py

```diff
@@ -109,3 +117,272 @@ def test_sdpa_tt_with_program_cache(device, b, nh, nkv, s, d, q_chunk_size, k_ch
+    )
+
+    padded_layer_len = nearest_n(start_idx, n=k_chunk_size)
+
+    # Test various sequence lengths
+    logger.debug(f"Testing with sequence length: {start_idx}")
+    logger.debug(f"Using chunk size: {k_chunk_size}")
+    logger.debug(f"Using padded layer length: {padded_layer_len}")
+    logger.debug(f"Using padded num heads: {padded_num_heads}")
+
+    attn_mask = torch.zeros((1, b, padded_num_heads, padded_layer_len))
+    # Assume all users are at same position
+    attn_mask[:, :, :, start_idx:] = torch.finfo(torch.float32).min
+
+    Q = torch.randn(1, b, padded_num_heads, d)
+
+    tt_Q = ttnn.as_tensor(Q, device=device, dtype=dtype, layout=ttnn.TILE_LAYOUT, memory_config=height_sharded_memcfg)
+
+    tt_attn_mask = ttnn.as_tensor(
+        attn_mask, device=device, dtype=dtype, layout=ttnn.TILE_LAYOUT, memory_config=dram_memcfg
+    )
+
+    tt_back = tt_lib.operations.primary.transformers.scaled_dot_product_attention(
+        tt_Q,
+        tt_K,
+        tt_V,
+        tt_attn_mask,
+        is_causal=False,
+        scale=scale,
+        program_config=program_config,
+        valid_seq_len=padded_layer_len,
+        compute_kernel_config=compute_kernel_config,
+        output_mem_config=height_sharded_memcfg,
+    )
+
+    tt_back = ttnn.to_torch(tt_back)
+    tt_back = tt_back[:, :, :nh, :]
+
+    Q_slice = Q[:, :, :nh, :].permute(1, 2, 0, 3)  # b, nh, 1, d
+    K_slice = K[:, :, :padded_layer_len, :].permute(1, 0, 2, 3)  # nh, b, S, d
+    V_slice = V[:, :, :padded_layer_len, :].permute(1, 0, 2, 3)  # nh, b, S, d
+    attn_mask_slice = attn_mask[:, :, :nh, :].permute(1, 2, 0, 3)  # b, nh, 1, S
+    expect = torch.nn.functional.scaled_dot_product_attention(
+        Q_slice, K_slice, V_slice, attn_mask_slice, scale=scale, is_causal=False
+    )  # b, nh, 1, d
+    expect = expect.squeeze().unsqueeze(0)
+
+    out_pass, out_pcc = comp_pcc(expect, tt_back, 0.99)
+
+    logger.debug(f"python vs pytorch: {out_pcc}")
+    assert out_pass
+
+
+@skip_for_grayskull("Unsupported in GS since L1 runs OOM with most configs")
+@pytest.mark.parametrize(
+    "dtype",
+    [tt_lib.tensor.DataType.BFLOAT8_B, tt_lib.tensor.DataType.BFLOAT16],
+    ids=["bfp8", "bf16"],
+)
+@pytest.mark.parametrize(
+    "b, nh, nkv, s, d",
+    ([16, 8, 1, 8192, 128],),  # Llama2-70B
+)
+def test_sdpa_decode_program_cache(device, b, nh, nkv, s, d, dtype, use_program_cache):
+    tt_lib.device.DisablePersistentKernelCache()
+
+    dummy_tensors = []
+    for _ in range(2):
+        dummy_tensors.append(
+            ttnn.as_tensor(
+                torch.zeros(32, 32),
+                device=device,
+                dtype=dtype,
+                layout=ttnn.TILE_LAYOUT,
+                memory_config=ttnn.types.MemoryConfig(
+                    ttnn.types.TensorMemoryLayout.INTERLEAVED, ttnn.types.BufferType.DRAM
+                ),
+            )
+        )
+        run_test_sdpa_decode_single_iter(device, b, nh, nkv, s, d, dtype)
```

Comment:
FYI, if you have sharded or L1 interleaved tensors, it's also not a bad idea to insert some tensor in L1. The buffer updates of sharded tensor is a separate API, so it might be worth testing.

Reply:
Added an L1 sharded tensor to test that my Q and output CB addresses get updated